File: Substitution Cipher.py
def encode(plainTxt, key):
    cipherTxt = ""
    lowerLetrs = "abcdefghijklmnopqrstuvwxyz"
    upperLetrs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    dict1 = {}
    
    for i in range(len(lowerLetrs)):
        dict1[lowerLetrs[i]] = key[i]
        dict1[upperLetrs[i]] = key[i].upper()
    
    for c in plainTxt:
        if c in dict1:
            cipherTxt += dict1[c]
        else:
            cipherTxt += c
    
    return cipherTxt

def decode(cipherTxt, key):
    decryptTxt = ""
    lowerLetrs = "abcdefghijklmnopqrstuvwxyz"
    upperLetrs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    dict2 = {}
    
    for i in range(len(lowerLetrs)):
        dict2[key[i]] = lowerLetrs[i]
        dict2[key[i].upper()] = upperLetrs[i]
    
    for c in cipherTxt:
        if c in dict2:
            decryptTxt += dict2[c]
        else:
            decryptTxt += c
    
    return decryptTxt

def isValidKey(key):
    # Check if key is 26 characters
    if len(key) != 26:
        return False
    
    # Check if key contains only alphabetic characters
    if not key.isalpha():
        return False
    
    # Check if all 26 characters are unique
    if len(set(key.lower())) != 26:
        return False
    
    return True

def main():
    while True:
        print("\nSubstitution Cipher Program")
        print("1. Encode Text")
        print("2. Decode Text")
        print("3. Exit")
        choice = int(input("Enter Your Choice: "))
        
        if choice == 1 or choice == 2:
            text = input("\nEnter The Text: ")
            key = input("Enter The Key (26 Unique Characters): ")
            
            if not isValidKey(key):
                print("Invalid Key! The Key Must Have 26 Unique Alphabetic Characters!")
                continue
            
            if choice == 1:
                print("\nEncoded Text:", encode(text, key))
            else:
                print("\nDecoded Text:", decode(text, key))
        elif choice == 3:
            print("Exiting Program...")
            break
        else:
            print("Invalid Choice! Please Choose Again!")

if __name__ == "__main__":
    main()

File: Substitution Cipher.cpp
#include <iostream>
#include <map>
#include <string>
#include <cctype>
using namespace std;

string encode(string plainTxt, string key) {
  string cipherTxt;
  string lowerLetrs = "abcdefghijklmnopqrstuvwxyz";
  string upperLetrs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  map < char, char > dict1;

  for (int i = 0; i < lowerLetrs.length(); i++) {
    dict1[lowerLetrs[i]] = key[i];
    dict1[upperLetrs[i]] = toupper(key[i]);
  }

  for (char & c: plainTxt) {
    if (dict1.find(c) != dict1.end()) {
      cipherTxt += dict1[c];
    } else {
      cipherTxt += c;
    }
  }

  return cipherTxt;
}

string decode(string cipherTxt, string key) {
  string decryptTxt;
  string lowerLetrs = "abcdefghijklmnopqrstuvwxyz";
  string upperLetrs = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

  map < char, char > dict2;

  for (int i = 0; i < lowerLetrs.length(); i++) {
    dict2[key[i]] = lowerLetrs[i];
    dict2[toupper(key[i])] = upperLetrs[i];
  }

  for (char & c: cipherTxt) {
    if (dict2.find(c) != dict2.end()) {
      decryptTxt += dict2[c];
    } else {
      decryptTxt += c;
    }
  }

  return decryptTxt;
}

int main() {
  int choice;
  string text, key;

  while (true) {
    cout << "\nSubstitution Cipher Program\n";
    cout << "1. Encode Text\n";
    cout << "2. Decode Text\n";
    cout << "3. Exit\n";
    cout << "Enter Your Choice: ";
    cin >> choice;

    if (choice == 1 || choice == 2) {
      cout << "\nEnter The Text: ";
      cin.ignore();
      getline(cin, text);
      cout << "Enter The Substitution Key (26 Characters For Alphabet): ";
      cin >> key;

      if (key.length() != 26) {
        cout << "Invalid Key Length! The Key Must Have 26 Characters!\n";
        continue;
      }

      if (choice == 1) {
        cout << "\nEncoded Text: " << encode(text, key) << endl;
      } else {
        cout << "\nDecoded Text: " << decode(text, key) << endl;
      }
    } else if (choice == 3) {
      cout << "Exiting Program...\n";
      break;
    } else {
      cout << "Invalid Choice! Please Choose Again!\n";
    }
  }
  return 0;
}
